File: typescript/dnd-character/dnd-character.ts
export class DnDCharacter {
  public strength: number;
  public dexterity: number;
  public constitution: number;
  public intelligence: number;
  public wisdom: number;
  public charisma: number;
  public hitpoints: number;

  public constructor() {
    this.strength = DnDCharacter.generateAbilityScore();
    this.dexterity = DnDCharacter.generateAbilityScore();
    this.constitution = DnDCharacter.generateAbilityScore();
    this.intelligence = DnDCharacter.generateAbilityScore();
    this.wisdom = DnDCharacter.generateAbilityScore();
    this.charisma = DnDCharacter.generateAbilityScore();
    this.hitpoints = 10 + DnDCharacter.getModifierFor(this.constitution);
  }

  public static generateAbilityScore(): number {
    const rolls = [rollDie(), rollDie(), rollDie(), rollDie()];
    rolls.sort();
    return rolls[1] + rolls[2] + rolls[3];
  }

  /** To calculate a modifier: subtract 10, divide by 2, then round down. */
  public static getModifierFor(abilityValue: number): number {
    return Math.floor((abilityValue - 10) / 2);
  }
}

function rollDie(): number {
  return Math.ceil(Math.random() * 6);
}

File: kotlin/dnd-character/src/main/kotlin/DndCharacter.kt
import kotlin.random.Random
import kotlin.random.nextInt

class DndCharacter {
    val strength: Int = ability()
    val dexterity: Int = ability()
    val constitution: Int = ability()
    val intelligence: Int = ability()
    val wisdom: Int = ability()
    val charisma: Int = ability()
    val hitpoints = 10 + modifier(constitution)

    companion object {
        fun diceRoll() = Random.nextInt(1..6)

        fun ability(): Int {
            val rolls = MutableList(4) { diceRoll() }
            rolls.sort()
            rolls.removeAt(0)
            return rolls.sum()
        }

        fun modifier(score: Int) = Math.floorDiv(score - 10, 2)
    }
}
